### PR TITLE
Add borders in Commit Form and set splitter width like in VS

### LINKED ISSUE #8712: Missing borders in FormCommit
## Current behaviour

#8557 has added borders to FormBrowse. This affected FormCommit, too, which is lacking borders now.

## Expected behaviour

At least add bottom borders to the list of unstaged files and to the patch view.

## Screenshots

current FormCommit

## Did this work in previous version of GitExtensions

former FormCommit

## Environment

- Git Extensions 33.33.33
- Build 468d7955a25faad218e16268ba2a56b07da5508c
- Git 2.27.0.windows.1 (recommended: 2.28.0 or later)
- Microsoft Windows NT 10.0.19042.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by ivangrek**
Fixes #8712

## Proposed changes

- Add bordes
- Set splitter width like in VS

### Before

image

### After

image

image

### Like
image

## Test methodology 

- Manually
- Visually

## Test environment(s) 

- Git Extensions 33.33.33
- Build aabe00bb81fba440c57170f601384418a882a06d (Dirty)
- Git 2.30.0.windows.2
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by mstv — COMMENTED** [review]
@ivangrek, please add screenshots of "before" in order to ease reviews.
Please consider maximized windows.
I expected only https://github.com/gitextensions/gitextensions/issues/8712#issuecomment-753682476.
@drewnoakes specifically removed several borders / sunken panels in the previous UI rework.

**Comment by mstv** [comment]
My 2ct: I don't think MS Visual Studio should be taken as the reference template.

**Comment by ivangrek** [comment]
> My 2ct: I don't think MS Visual Studio should be taken as the reference template.

I agree, it doesn't look as modern as we can see in many other applications.

But I have these thoughts.
Now it is more important to come to an unambiguous interface in all dialogues. The simplest and most understandable will be to take as a basis what is already there, used and understandable. We could consider this as an intermediate goal for the transition to a more modern design. There are many design nuances and we can get bogged down in that.

**Comment by ivangrek** [comment]
> I expected only #8712 (comment).

Anyway, if it causes such resistance, I can suggest using the experimental setting to enable this mode.
Otherwise, we can roll everything back, since the movement in this direction takes too much time and effort.

**Comment by drewnoakes** [comment]
I removed lots of borders, mostly trying to simplify things. We had so many 1px borders, nested, with inconsistent padding. Some places they would have no padding, leading to nested controls producing visible 2px or 3px borders. Removing everything seemed sensible. I'm not opposed to reintroducing some borders in an intentional and controlled manner.

Consider too that we have many more dialogs that would benefit from consistent styling (push, pull, rebase...). Whatever style design evolves here should be applicable across the board.

Keep up the great work @ivangrek!

**Comment by RussKie** [comment]
There's a work item that track consistency of dialogs layouts.
Few dialogs have been modernised, but there's still a long way to go.

**Comment by ivangrek** [comment]
Ref to #8745.

Saving/restoring the splitter position is implemented incorrectly, which leads to breakdown of tests.
The restoration of the position should take place on Form_Show event, and not before that.

**Comment by pmiossec** [comment]
Do you think it could be a good idea to remove the margin and the 2 borders between the unstaged and staged files control?

image

Because I don't know why the stage/unstage toolbar belong more to the bottom control than the top control....

**Comment by ivangrek** [comment]
> Do you think it could be a good idea to remove the margin and the 2 borders between the unstaged and staged files control?

That's a good question. Now it is so implemented. I do not want to dive into the details of each, but focused on the main idea of bringing to an unambiguous form. As far as this raises other issues as well.

Perhaps in the current situation it will be easier to transfer buttons to another panel.
But on the other hand, these buttons are in the area of responsibility of the lower panel, while the upper one is for viewing.

I could suggest moving towards a similar design as the left panel, but we're not ready for that yet. 😄

**Comment by RussKie** [comment]
My 20c - I _do_ like the new main window layout, to me personally it feels significantly more modern and closer to the current UI/UX paradigms of Windows and Office than the current borderless/marginless layout.

With that I _do_ like the proposed layout 👍

**Comment by RussKie** [comment]
> image
> 
> Because I don't know why the stage/unstage toolbar belong more to the bottom control than the top control....

Good observation @pmiossec!
I don't have an answer here either, and I'd split this into a separate UI design discussion. And I throw a more radical idea into the mix - "Reset all changes" and "Reset unstaged" buttons probably belong more to the same panel as "Stage" and "Unstage" buttons :D

**Comment by ivangrek** [comment]
Can we merge this PR?

**Comment by mstv** [comment]
> Can we merge this PR?

Not yet because of just disabling broken splitter positions in https://github.com/gitextensions/gitextensions/pull/8849/commits/d83a7939c9d93a9add5a313766cc782a96ac49a3. Don't forget #8745, too!

BTW: Much less noisy:
```
#if false
    disabled code
#endif
```

**Review by RussKie — COMMENTED** [review]
What's the deal with the commented out code?

**Comment by ivangrek** [comment]
> What's the deal with the commented out code?

Splitter position is work incorrectly. Splitters affect the position of each other in sequential recovery.

**Comment by RussKie** [comment]
The commit message doesn't explain this, there is brief mention of this on
GitHub. Please update the commit message with an explanation to make it
easier to appreciate the change.
Also I'd follow @mstv suggestion to use #if..#endif or /* */ to reduce the
noise.

So we're fixing one thing and breaking another... Are you working on
getting the splitters working? We can't take this into v3.5 without them.

**Comment by ivangrek** [comment]
> Also I'd follow @mstv suggestion to use #if..#endif or /* */ to reduce the
noise.

In general, or instead of experimental settings?

**Comment by ivangrek** [comment]
> Are you working on getting the splitters working?

I am researching what can be done, but do not see a solution yet.

**Comment by RussKie** [comment]
> > Also I'd follow @mstv suggestion to use #if..#endif or /* */ to reduce the
> > noise.
> 
> In general, or instead of experimental settings?

In this PR :)

**Comment by RussKie** [comment]
> > Are you working on getting the splitters working?
> 
> I am researching what can be done, but do not see a solution yet.

I'll try to find some time this week to pull it down and have a look.

**Comment by ivangrek** [comment]
> In this PR :)

Not quite sure how to enable / disable this?

File: GitUI/SplitterManager.cs [hunk 1]
```diff
@@ -29,18 +29,18 @@ public void AddSplitter(SplitContainer splitter, string settingName, int? defaul
...
         public void RestoreSplitters()
         {
-            foreach (var splitter in _splitters)
-            {
-                splitter.RestoreFromSettings(_settings);
-            }
+            ////foreach (var splitter in _splitters)
+            ////{
+            ////    splitter.RestoreFromSettings(_settings);
+            ////}
```
Comment (RussKie) [inline GitUI/SplitterManager.cs:35]:
```suggestion
            /*
             Disabled as splitters have become unstable. Refer to #123456 for more details.
             
            foreach (var splitter in _splitters)
            {
                splitter.RestoreFromSettings(_settings);
            }
            */
```

Please create a new issue detailing the problem, and reference it here and everywhere else where the code is commented.

Comment (ivangrek) [reply GitUI/SplitterManager.cs:35]: #8745

**Comment by RussKie** [comment]
I've run this branch locally, and here are my observations:

1. I can see a small vertical creep of the horizontal splitter (the orange arrow). I couldn't see any splitter instabilities in the `FormBrowse`. I would classify this as minor mildly annoying issue.
2. Visual artifacts marked by the red and cyan colours, these are significantly more concerning. The left panel's top panel exhibits the same issue btw.

image

**Comment by RussKie** [comment]
Looks like we're rendering the toolbar with rounded corners - @NikolayXHD any ideas?
image

**Comment by ivangrek** [comment]
I think Professional Renderer try to render shadows.

**Comment by RussKie** [comment]
The doubled right border comes from `FileStatusList` from `FileStatusListView`.
image

**Comment by RussKie** [comment]
> 2\. Visual artifacts marked by the red and cyan colours, these are significantly more concerning. The left panel's top panel exhibits the same issue btw.

This looks like a theming-related issue:

| Default | Dark |
--|--
image | image

In the default (system) theme the column vertical line is of a faint light blue colour. In the dark theme - the column line's colour is significantly more prominent, and looks the same as the colour as the border.


The toolbar rendering issue only appears in the "themed" scenarios too.

**Comment by RussKie** [comment]
I propose we toned down the borders a little:
```diff
-            splitLeft.Panel1.BackColor = Color.LightGray.AdaptBackColor();
-            splitLeft.Panel2.BackColor = Color.LightGray.AdaptBackColor();
-            splitRight.Panel1.BackColor = Color.LightGray.AdaptBackColor();
-            splitRight.Panel2.BackColor = Color.LightGray.AdaptBackColor();
+            Color borderColor = KnownColor.ControlLightLight.MakeBackgroundDarkerBy(0.08);
+            splitLeft.Panel1.BackColor = borderColor;
+            splitLeft.Panel2.BackColor = borderColor;
+            splitRight.Panel1.BackColor = borderColor;
+            splitRight.Panel2.BackColor = borderColor;
```

<details>
<summary>Current proposal</summary>

* Default
image
* Dark
image
* Dark silver
image | 


</details>

<details>
<summary>New proposal</summary>

* Default
image
* Dark
image
* Dark silver
image



</details>

<details>
<summary>Bonus</summary>





Applied to the main window, it looks like this:
* Default
image
* Dark
image
* Dark silver
image

</details>

**Comment by ivangrek** [comment]
> I propose we tone down the borders a little

👍 but suggest ` #dddddd`

**Comment by ivangrek** [comment]
Why tests locally and on AppVeyor may differ?

File: GitUI/SplitterManager.cs [hunk 2]
```diff
@@ -95,7 +95,8 @@ public void RestoreFromSettings(ISettingsSource settings)
                                 break;
                             case FixedPanel.Panel2:
                                 int panel2PrevSize = DpiUtil.Scale(prevSize, prevDpi) - DpiUtil.Scale(prevDistance, prevDpi);
-                                SetSplitterDistance(SplitterSize - panel2PrevSize + splitterWidth);
+                                const int paddingOffset = 2; // Refer to FormCommit.ctor+WorkaroundPaddingIncreaseBug
```
Comment (RussKie) [inline GitUI/SplitterManager.cs:98]: @ivangrek this makes the splitter stable at the runtime again, however the tests on the CI are still unstable

Comment (ivangrek) [reply GitUI/SplitterManager.cs:100]: I don't notice any shifts for the "Commit Form", but sometimes there are changes for the main form.

**Comment by RussKie** [comment]
In essence the change is happening for
splitcontainers with FixedPanel=FixedPanel.Panel2.

How to repro:
1. Open FormCommit
2. Take a screenshot1
3. Close FormCommit
4. Repeat 1-3
5. Compare screenshot1 with screenshot2

You can also attach a debugger and set a breakpoint at SetSplitterDistance
method and see the value changing for "splitRight_" component.

**Comment by ivangrek** [comment]
I marked the points with a pencil on the monitor and just reopen the window and see if there are any shifts. 😄

**Comment by RussKie** [comment]
Not going to work. Use a permanent marker 😁

**Pull request merged**